firmware/Core/Src/flash_bank/flash_bank.c

```diff
@@ -0,0 +1,124 @@
+{
+    HAL_StatusTypeDef status;
+    uint32_t endAddress = address + length;
+
+    status = HAL_FLASH_Unlock();
+
+    if (status != HAL_OK)
+    {
+        return 1;
+    }
+
+    // Clear all FLASH flags before starting the operation
+    __HAL_FLASH_CLEAR_FLAG(FLASH_FLAG_ALL_ERRORS);
+
+    // Write data to flash
+    for (uint32_t currentAddress = address; currentAddress < endAddress; currentAddress += 8)
+    {
+        // currentAddress - address is the number of bytes already written
+        // const uint64_t double_word = *(uint64_t *)(data + (currentAddress - address));
+        uint8_t data_to_write[8] = {0};
+        memcpy(data_to_write, data + (currentAddress - address), 8);
+        const uint64_t double_word = *(uint64_t *)(data_to_write);
+        // Write the word to flash
+        status = HAL_FLASH_Program(FLASH_TYPEPROGRAM_DOUBLEWORD, currentAddress, double_word);
+        if (status != HAL_OK)
+        {
+            // If the program operation fails, break the loop
+            break;
+        }
+    }
+
+    // Lock the Flash to disable the flash control register access
+    HAL_FLASH_Lock();
+
+    return status;
+}
+
+/**
+ * Reads data from the flash memory.
+ *
+ * This function takes in an address, a pointer to a buffer to store the read data, and the length of the data to be read.
+ * It checks if the address is within the valid range of the flash memory and returns an error if it's not.
+ * If the address is valid, it reads the data from the flash memory and stores it in the provided buffer.
+ *
+ * @param address The starting address in the flash memory where the data will be read from.
+ * @param buffer A pointer to the buffer where the read data will be stored.
+ * @param length The length of the data to be read.
+ *
+ * @return 0 if the read operation is successful, 1 if the address is out of range.
+ */
+uint8_t Flash_Bank_Read(uint32_t address, uint8_t *buffer, uint32_t length)
+{
+    // Check if the address is within the valid range of the flash memory
+    if (address < Flash_Partitions_FLASH_BANK2 || (address + length) > (FLASH_BANK2_END))
+    {
+        return 1; // Return error if address is out of range
+    }
+
+    for (uint32_t i = 0; i < length - 1; i++)
+    {
+        buffer[i] = *(uint8_t *)(address + i);
+    }
+    buffer[length] = '\0';
+
+    return 0; // Return success
+}
+
+/**
+ * Erases a specified number of pages from the flash bank.
+ *
+ * This function takes in the starting page to erase and the number of pages to erase.
+ * It clears any existing flash errors, unlocks the flash, and then erases the specified pages.
+ * If the erase operation is successful, it returns 0. Otherwise, it returns the page error.
+ *
+ * @param start_page_erase The starting page to erase.
+ * @param number_of_pages_to_erase The number of pages to erase.
+ *
+ * @return 0 if the erase operation is successful, page error otherwise.
```

Comment:
What is "page error"?

Reply:
I left a comment explaining what page error is